cmd/api/src/api/v2/saved_queries.go

```diff
@@ -100,3 +102,32 @@ func (s Resources) CreateSavedQuery(response http.ResponseWriter, request *http.
 		api.WriteBasicResponse(request.Context(), savedQuery, http.StatusCreated, response)
 	}
 }
+
+func (s Resources) DeleteSavedQuery(response http.ResponseWriter, request *http.Request) {
+	var (
+		rawSavedQueryID = mux.Vars(request)[api.URIPathVariableSavedQueryID]
+		userID          = ctx2.FromRequest(request).AuthCtx.Session.UserID
+	)
+
+	if savedQueryID, err := strconv.Atoi(rawSavedQueryID); err != nil {
+		api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, api.ErrorResponseDetailsIDMalformed, request), response)
+	} else if savedQueries, _, err := s.DB.ListSavedQueries(userID, "", model.SQLFilter{}, 0, 10000); err != nil {
+		api.HandleDatabaseError(request, response, err)
+	} else {
+		// ensure that the query ID specified belongs to the current user
+		savedQueryIDMatched := false
+		for _, savedQuery := range savedQueries {
+			if savedQuery.ID == int64(savedQueryID) {
+				savedQueryIDMatched = true
+				break
+			}
+		}
+
+		if !savedQueryIDMatched {
+			api.WriteErrorResponse(request.Context(), api.BuildErrorResponse(http.StatusBadRequest, "invalid saved_query_id supplied", request), response)
+		} else {
```

Comment:
I feel like we could just use SQL to manage this complexity rather than pulling "all" (technically only the first 10000, which is likely to be all in most real world scenarios, but importantly isn't guaranteed to be all) queries for the current user and manually sifting. This would mean adding another DB method for this purpose, but it'll be a lot cleaner. Something like:

```go
func (s *BloodhoundDB) UserOwnsSavedQuery(userID uuid.UUID, id int) (bool, error) {
	var count int64
	if result := s.db.Where("user_id = ?", userID).Where("id = ?", id).Find(model.SavedQuery{}).Count(&count); result.Error != nil {
		return false, result.Error
	} else {
		return count == 1, nil
	}
}
```

You can then do error checking and a bool check, then proceed with the delete without any additional application level handling.

Reply:
The accepted solution is going to be similar to this, but much better than my quick scratch fix. One thing we want to handle with the equivalent of `UserOwnsSavedQuery()` is to check the error here at the API level and have a special case for `errors.Is(err, database.ErrNotFound)`, returning a 404 to the user to let the know the record has already been deleted (or doesn't exist)